### PR TITLE
Fix up/final

### PR DESCRIPTION
**Pull request by tjj2017**
Handles private and incomplete declarations whose full type declaration uses type declared between the partial declaration and its completion.

This feature was required to address issue #194.

### CODE REVIEW

File: testsuite/gnat2goto/tests/loop_array_range/loop_array_range.adb [hunk 1]
```diff
@@ -8,5 +8,5 @@ begin
       Sum := Sum + My_Arr(I);
    end loop;
 
-   pragma Assert(Sum=6);
+   pragma Assert(Sum=10);
```
Comment (martin-cs) [inline testsuite/gnat2goto/tests/loop_array_range/loop_array_range.adb:11]: Why change this?  Why not have both?  A test can still pass if CBMC says `VERIFICATION FAILED`.  It is better to have a mix of tests TBH.

File: testsuite/gnat2goto/tests/range_third/test.opt [hunk 2]
```diff
@@ -0,0 +1,2 @@
+ALL XFAIL Doesn't crash anymore but cbmc does not perform range assertion checks 
+
```
Comment (martin-cs) [inline testsuite/gnat2goto/tests/range_third/test.opt:2]: @jeremycrisp worth keeping in mind.

**Review by martin-cs — APPROVED** [review]
A couple of questions about tests but basically, yes.

File: testsuite/gnat2goto/tests/range_third/test.opt (hunk 2 above)
Comment (tjj2017) [reply testsuite/gnat2goto/tests/range_third/test.opt:2]: I felt the purpose of the test was to check the loop and therefore felt a positive result was required.  I have added a negative test in which the assertion fails.

**Review by martin-cs — APPROVED** [review]
Great; thank you.

**Comment by tjj2017** [comment]
Don't merge yet I have just located some debug code I have left in.  I will push again.

**Pull request merged**